### PR TITLE
Thimble issue1763 - Adding whitespace indicators to the editor

### PR DESCRIPTION
**Pull request by tonypark0403**
Added whitespace by default

### CODE REVIEW

**Comment by tonypark0403** [comment]
This issue comes from https://github.com/mozilla/thimble.mozilla.org/issues/1763.
And I added the whitespace indicator by default.

File: Gruntfile.js [hunk 1]
```diff
@@ -522,4 +523,4 @@ module.exports = function (grunt) {
 
     // Default task.
     grunt.registerTask('default', ['test']);
-};
+};
```
Comment (humphd) [inline Gruntfile.js]: Here's an example of what we discussed in class about not having an end-of-file line break.  Please revert this change back.

File: package.json [hunk 2]
```diff
@@ -70,4 +70,4 @@
         "start": "http-server -p 8000 --cors"
     },
     "license": "MIT"
-}
+}
```
Comment (humphd) [inline package.json]: Same issue here, please revert.

File: src/config.json [hunk 3]
```diff
@@ -37,59 +37,50 @@
...
-        "nunjucks": "2.3.0",
+        "nunjucks": "^3.0.0",
         "properties-parser": "0.3.1",
         "request": "^2.69.0",
-        "requirejs": "2.1.22",
-        "sw-precache": "^3.1.1",
-        "tar-fs": "^1.11.1"
+        "requirejs": "^2.3.3",
+        "sw-precache": "^5.0.0",
+        "tar-fs": "^1.15.1"
```
Comment (humphd) [inline src/config.json]: All of these changes need to be reverted--it looks like you need to rebase.

File: src/utils/BrambleExtensionLoader.js [hunk 4]
```diff
@@ -34,6 +34,7 @@ define(function (require, exports, module) {
         "bramble",
         "Autosave",
         "brackets-paste-and-indent",
+        "brackets-show-whitespace",
```
Comment (humphd) [inline src/utils/BrambleExtensionLoader.js]: I have serious misgivings about making this a default that we load.  @flukeout this needs to be tested by you, and I also want to do performance testing to make sure it doesn't bog down the editor.

**Comment by humphd** [comment]
@flukeout can you test/review this?  Especially pay attention to perf.

**Comment by flukeout** [comment]
Will do!

**Comment by flukeout** [comment]
@tonypark0403 I'm not able to see the white space indicators in the Editor. I pulled your changes into brackets and then ran ``npm run build``. In the URL i also added ``?enableExtensions=brackets-show-whitespace`` - is there a step I am missing to see these changes?

cc @humphd - did you get it running?

**Comment by tonypark0403** [comment]
Yes, it is.
Since it is not decided to provide the feature by default, I deleted the line.
You can refer to this address, https://github.com/mozilla/brackets/pull/630/files/ddbe293cb1c1e36f04535fad130d4e55d8510098

In the file, src/utils/BrambleExtensionLoader.js, you should add "brackets-show-whitespace" in the 36 line.

**Comment by flukeout** [comment]
Thanks @tonypark0403 - trying it out now.

**Comment by flukeout** [comment]
Hi @tonypark0403 - I added the extension in ``src/utils/BrambleExtensionLoader.js``...

image

...but still don't see any whitespaces.

image

**Comment by tonypark0403** [comment]
I added the tmpl files and applied it. Try it.

**Comment by flukeout** [comment]
Will try now - thank you.

**Comment by flukeout** [comment]
This works great @tonypark0403 - nice work! Finally got everything running. Ultimately, I think this should be an option in the gear menu dropdown and should be turned on by default, like this...

image

However, to get there I think we should...

* First, land this pull request but disable the extension by default
* Second, start a new issue related to adding it to the menu

@tonypark0403 - can you change it so that the extension is NOT enabled by default?

Then we can do a final review of the code. Does that sound good?

File: src/extensions/default/brackets-show-whitespace/main.js [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        stylesTemplate     = require("text!styles/whitespace-colors-css.tmpl"),
+        Strings            = require("strings");
+
+    
+    // --- Settings ---
+    
+    var commandID          = "denniskehrig.ShowWhitespace.toggle";
+    var preferencesID      = "denniskehrig.ShowWhitespace";
+    var defaultPreferences = {
+        enabled: true,
```
Comment (humphd) [inline src/extensions/default/brackets-show-whitespace/main.js]: Let's change this to `false` by default, so that we have to flip a pref in order to enable it.  Then you can load it all the time.

**Comment by humphd** [comment]
I've updated https://github.com/mozilla/thimble.mozilla.org/issues/1763#issuecomment-287172964 to be about the UI toggle.

**Comment by humphd** [comment]
@tonypark0403 when you add new commits and it's ready for review again, you should @ mention the person who is going to do the review so they know.  Otherwise we lose track.  I'll review this now.

File: src/extensions/default/brackets-show-whitespace/styles/whitespace-colors-css.tmpl [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/*
```
Comment (humphd) [inline src/extensions/default/brackets-show-whitespace/styles/whitespace-colors-css.tmpl]: @flukeout do you know what a `.tmpl` file is with CSS?  I've never seen this before.

**Comment by humphd** [comment]
Tested this locally, it works great.  Nice work @tonypark0403.  Now you can do the Thimble side and add UI to enable/disable it.

**Comment by humphd** [comment]
@tonypark0403 so here is my suggestion for you:

* 0.2 is this PR that adds the extension to Brackets, but it's default to `enabled=false`
* 0.3 is the part in Brackets that adds `bramble.enableWhiteSpace()` and `bramble.disableWhiteSpace()` which is similar to [this](https://github.com/mozilla/brackets/commit/b524178098f17ccf129268daf4664493916835cb)
* 0.4 is the Thimble UI in the menu that calls what you wrote in 0.3 and would be done in https://github.com/mozilla/thimble.mozilla.org/issues/1763
